docs/cancel-and-close-unpaid-orders-after-two-days/script.liquid

```diff
@@ -56,143 +28,266 @@
 {% assign period_s = period_unit | times: options.period_to_wait_before_checking_each_order__number_required %}
 {% assign maximum_date_s = "now" | date: "%s" | minus: period_s %}
 
-{% for order in orders %}
-  {% assign order_tags = order.tags | split: ", " %}
+{% if options.test_mode__boolean %}
+  {% assign test_mode_summaries = array %}
+{% endif %}
 
+{% capture orders_query %}
+  financial_status:pending
   {% if options.only_process_orders_having_this_tag != blank %}
-    {% unless order_tags contains options.only_process_orders_having_this_tag %}
-      {% continue %}
-    {% endunless %}
+    AND tag:{{ options.only_process_orders_having_this_tag | strip | json }}
   {% endif %}
-
   {% if options.ignore_orders_having_this_tag != blank %}
-    {% if order_tags contains options.ignore_orders_having_this_tag %}
-      {% continue %}
-    {% endif %}
+    AND -tag:{{ options.ignore_orders_having_this_tag | strip | json }}
   {% endif %}
+{% endcapture %}
 
-  {% assign processed_at_s = order.processed_at | default: order.created_at | date: "%s" | times: 1 %}
+{% assign orders_query = orders_query | strip_newlines %}
 
-  {% if processed_at_s > maximum_date_s %}
-    {% continue %}
-  {% endif %}
+{% log orders_query: orders_query %}
 
-  {% assign summary = hash %}
-  {% assign summary["order_name"] = order.name %}
-  {% assign summary["order_id"] = order.id %}
-  {% assign summary["order_processed_at"] = processed_at_s | date: "%FT%T%:z" %}
-  {% assign summary["threshold_processed_at"] = maximum_date_s | date: "%FT%T%:z" %}
+{% assign cursor = nil %}
 
-  {% if order.cancelled_at == blank %}
-    {% assign summary["qualifies_for_cancellation"] = true %}
-  {% else %}
-    {% assign summary["qualifies_for_cancellation"] = false %}
-  {% endif %}
+{% for n in (0..1000) %}
+  {% capture query %}
+    query {
+      orders(
+        first: 5
+        after: {{ cursor | json }}
+        query: {{ orders_query | json }}
+      ) {
+        pageInfo {
+          hasNextPage
+        }
+        edges {
+          cursor
+          node {
+            id
+            name
+            createdAt
+            processedAt
+            cancelledAt
+            closedAt
+            tags
+            lineItems(first: 25) {
```

Comment:
Rather than dealing with a limited set of line items (and potentially going the direction of a bulk operation to sidestep that concern), rework this using `Order.suggestedRefund(suggestFullRefund: true)` ([docs](https://shopify.dev/docs/admin-api/graphql/reference/orders/order?api%5Bversion%5D=2020-10#suggestedrefund-2020-10)). That'll give us an (unpaginated) set of refund line item suggestions, which (bonus!) come with location and transaction suggestions, too.

Also this should drop the query cost. 🙌

Reply:
Ooh, nice. Curiously though, the **restockType** suggestions seem to always be CANCEL and **restockable** is _true_ for either tracked or untracked inventory, which would imply there is no need to differentiate between RETURN/NO_RESTOCK. I presume there might be additional reasons to do so that I haven't ascertained in my testing, so let me know if I should just keep the logic as is (but subbing out for the restockable check of course). 

Also, **suggestedTransactions** appears to always be empty for pending orders, which makes sense as no payment has been authorized or received, so there shouldn't be any payment to refund. I also tested paying for an order and then editing it to add an additional item, and this partially paid edited order does not fall within the ```financial_status:pending``` filter. So, I'll stick with the order.transactions.first logic to handle the VOID.

Reply:
Restocking: huh, good sleuthing. Keep the logic as-is, then. :)

Transactions logic: cool, makes sense. Thanks for testing; I trust that call. :)